// Copyright (c) 2011-2012 Stanford University
//
// Permission to use, copy, modify, and distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

package LogCabin.ProtoBuf.InternalLog;

/**
 * \file
 * This file contains the formats for the log entries that make up each
 * server's internal log. This internal log is used to describe the other logs
 * in the system.
 */


message LogEntry {
    enum Type {
        METADATA_TYPE = 10;
        DECLARE_LOG_TYPE = 11;

        // Seriously, don't use this type in production code.
        ILLEGAL_TYPE_FOR_TESTING_PURPOSES = 666;
    }

    // Identifies which field is filled in.
    required Type type = 1;

    // Tags 2 through 9 are reserved for future use.

    // One of the following will be filled in.
    optional Metadata metadata = 10;
    optional DeclareLog declare_log = 11;
}

message Metadata {
    // A unique ID for the contents of this storage.
    // This should be a big random string.
    required string uuid = 1;
}

message DeclareLog {
    // The name for the log being declared.
    // This entry's EntryId will be its LogId.
    required string log_name = 1;
}
